### PR TITLE
Blackbox Exporter guide

### PR DESCRIPTION
**Pull request by lucperkins**
cc @brian-brazil @juliusv @RichiH 

This is now ready for review.

### CODE REVIEW

**Comment by brian-brazil** [comment]
I likely won't get to this until Tuesday. One initial comment is that building out a custom Python app including a full framework is unnecessary. Use what you have already, such as the node exporter.

File: content/docs/guides/blackbox-exporter.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+```yaml
+modules:
+  http_2xx:
+    prober: http
+    timeout: 5s
+    http:
+      method: GET
+```
+
+With this configuration, the Blackbox Exporter will use the built-in `http_2xx` module, which periodically probes HTTP endpoints (the interval is set in your Prometheus configuration).
```
Comment (juliusv) [inline content/docs/guides/blackbox-exporter.md]: "the built-in `http_2xx` module" -> The name `http_2xx` is actually arbitary and user-chosen in the configuration, so there's no built-in module `http_2xx`. It would be more correct to say that this configuration defines a `http_2xx` module that uses the built-in `http` prober.

File: content/docs/guides/node-exporter.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ NOTE: While the Prometheus Node Exporter is for *nix systems, there is a [WMI ex
 
 ## Installing and running the Node Exporter
 
-The Prometheus Node Exporter is a single static binary that you can install [via tarball](#tarball-installation). Once you've downloaded it from the Prometheus [downloads page](/download#node_exporter) extract it, and run it:
+The Prometheus Node Exporter is a single static binary that you can install via tarball. Once you've downloaded it from the Prometheus [downloads page](/download#node_exporter) extract it, and run it:
```
Comment (juliusv) [inline content/docs/guides/node-exporter.md]: It seems like this is an unrelated change that made it into this PR?

**Comment by juliusv** [comment]
> One initial comment is that building out a custom Python app including a full framework is unnecessary. Use what you have already, such as the node exporter.

- If we just want to probe for a 200 response status code, I agree.
- Another option would be to make the example more complex and return a specific string that we probe for in the HTTP response (you know, like checking that some API call contains expected fields) - that would be a better reason for introducing custom code, but it would also make the example less basic.
- As a third option (and one I generally like), we could omit running the probe target altogether and just probe public targets like `https//google.com`, `https://prometheus.io/`, and `https://twitter.com`.

**Comment by RichiH** [comment]
A fourth option, which is probably out of scope here, would be to extend blackbox_exporter to tackle more complex checks.

More within scope would be the keep a truly basic howto (Julius #3) and extend to a more complex custom API checker thing later (Julius #2)

File: content/docs/guides/node-exporter.md (hunk 2 above)
Comment (lucperkins) [reply content/docs/guides/node-exporter.md]: It's an unrelated change but I felt it was small enough to include here (it fixes a broken link). I can put it in a separate PR if you'd like.

**Comment by lucperkins** [comment]
@juliusv I do like the simplicity of probing a public target like that. I'll push a new draft that takes that approach and we'll see what Richie and Brian have to say 😄

**Comment by brian-brazil** [comment]
The risk with public targets is that the content may change, or we could be limited by anti-abuse stuff. You will also run into v6 issues on all those endpoints. Something local is safest.

File: content/docs/guides/blackbox-exporter.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+```yaml
+scrape_configs:
+- job_name: blackbox
+  metrics_path: /probe
+  scrape_interval: 5s
+  params:
+    module: [http_2xx]
+  static_configs:
+  - targets:
+    - localhost:2112/health
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: It'd be safest to quote this, it's less likely the user will have issues later as they adjust it

File: content/docs/guides/blackbox-exporter.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  relabel_configs:
+  - source_labels: [__address__]
+    target_label: __param_target
+  - source_labels: [__param_target]
+    target_label: instance
+  - target_label: __address__
+    replacement: localhost:9115
+```
+
+This instructs Prometheus to scrape the Blackbox Exporter's `/probe` endpoint and passes the `targets` list, which includes the Python web server running on port 2112, to the Blackbox Exporter. With this configuration in place, start Prometheus:
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: target is a single item

File: content/docs/guides/blackbox-exporter.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+```shell
+./prometheus
+```
+
+## Installing, configuring, and running the Blackbox Exporter
+
+Like Prometheus, the Blackbox Exporter is a single static binary that you can install via tarball. [Download the latest release](/download#blackbox_exporter) for your platform and untar it:
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: I'd setup and play with the blackbox exporter first, then talk about Prometheus

File: content/docs/guides/blackbox-exporter.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+```bash
+wget https://github.com/prometheus/blackbox_exporter/releases/download/v*/blackbox_exporter-*.*-amd64.tar.gz
+tar xvfz blackbox_exporter-*.*-amd64.tar.gz
+cd blackbox_exporter-*.*-amd64
+```
+
+The untarred folder contains a `blackbox.yml` file for configuring the exporter. Delete the current contents and replace it with this:
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: This is already the default, what is the benefit of the user changing this?

File: content/docs/guides/blackbox-exporter.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  http_2xx:
+    prober: http
+    timeout: 5s
+    http:
+      method: GET
+```
+
+With this configuration, the Blackbox Exporter will use the built-in `http_2xx` module, which periodically probes HTTP endpoints (the interval is set in your Prometheus configuration).
+
+NOTE: In the Blackbox Exporter configuration, you don't need to specify host or port information for the specific endpoints that you want to probe, only which modules you'd like to use. Host and port information is specifed in the [Prometheus configuration](#prometheus-configuration).
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: They're targets, not host:port. HTTP can have more, icmp is just host.

File: content/docs/guides/blackbox-exporter.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+Navigate to http://localhost:9115 in your browser to see the Blackbox Exporter's web interface. Under the **Recent Probes** section you'll see results from the exporter's most recent probes as a table. You should see entries like this:
+
+Module | Target | Result | Debug
+:------|:-------|:-------|:-----
+`http_2xx` | `localhost:2112/health` | Success | Logs
+`http_2xx` | `localhost:2112/health` | Success | Logs
+`http_2xx` | `localhost:2112/health` | Success | Logs
+
+In the Prometheus [expression browser](/docs/visualization/browser), accessible at http://localhost:9090/graph, you can a variety of Blackbox-Exporter-specific metrics, which have the `probe_` prefix, for example, the [`probe_success{instance="localhost:2112/health",job="blackbox"}`](http://localhost:9090/graph?g0.range_input=1h&g0.expr=probe_success{instance%3D%localhost%3A2112%2Fhealth%22}), which indicates whether the most recent probe was successful.
```
Comment (brian-brazil) [inline content/docs/guides/blackbox-exporter.md]: You need to talk more about probe_success, many users miss the importance of this

Maybe setup an alert?


I'd also mention the debug URL parameter.

**Comment by roidelapluie** [comment]
@lucperkins Are you still interested by this?

**Comment by lucperkins** [comment]
@roidelapluie I'm no longer with the CNCF, so it's unlikely that I will ever complete this.

**Comment by roidelapluie** [comment]
@lucperkins Can I take over your branch and take it to the finish line?

**Comment by lucperkins** [comment]
@roidelapluie Absolutely! You have my full endorsement 💯
